// @pothos/plugin-errors/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    default: ()=>_default,
    capitalize: ()=>capitalize,
    defaultGetResultName: ()=>defaultGetResultName,
    defaultGetUnionName: ()=>defaultGetUnionName,
    unwrapError: ()=>unwrapError,
    PothosErrorsPlugin: ()=>PothosErrorsPlugin
});
require("./global-types");
const _core = /*#__PURE__*/ _interopRequireWildcard(require("@pothos/core"));
_exportStar(require("./types"), exports);
function _exportStar(from, to) {
    Object.keys(from).forEach(function(k) {
        if (k !== "default" && !Object.prototype.hasOwnProperty.call(to, k)) Object.defineProperty(to, k, {
            enumerable: true,
            get: function() {
                return from[k];
            }
        });
    });
    return from;
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interopRequireWildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {};
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const pluginName = 'errors';
const _default = pluginName;
function capitalize(s) {
    return `${s.slice(0, 1).toUpperCase()}${s.slice(1)}`;
}
const defaultGetResultName = ({ parentTypeName , fieldName  })=>`${parentTypeName}${fieldName}Success`;
const defaultGetUnionName = ({ parentTypeName , fieldName  })=>`${parentTypeName}${fieldName}Result`;
const unwrapError = Symbol.for('Pothos.unwrapErrors');
function createErrorProxy(target, ref, state) {
    return new Proxy(target, {
        get (err, val, receiver) {
            if (val === unwrapError) {
                return ()=>{
                    // eslint-disable-next-line no-param-reassign
                    state.wrapped = false;
                };
            }
            if (val === _core.typeBrandKey) {
                return ref;
            }
            return Reflect.get(err, val, receiver);
        },
        getPrototypeOf (err) {
            const proto = Reflect.getPrototypeOf(err);
            if (!state.wrapped || !proto) {
                return proto;
            }
            return createErrorProxy(proto, ref, state);
        }
    });
}
const errorTypeMap = new WeakMap();
class PothosErrorsPlugin extends _core.BasePlugin {
    wrapIsTypeOf(isTypeOf, config) {
        if (isTypeOf) {
            return (parent, context, info)=>{
                if (typeof parent === 'object' && parent) {
                    var _parent_unwrapError;
                    (_parent_unwrapError = parent[unwrapError]) === null || _parent_unwrapError === void 0 ? void 0 : _parent_unwrapError.call(parent);
                }
                return isTypeOf(parent, context, info);
            };
        }
        return isTypeOf;
    }
    onOutputFieldConfig(fieldConfig) {
        const errorOptions = fieldConfig.pothosOptions.errors;
        const errorBuilderOptions = this.builder.options.errorOptions;
        if (!errorOptions) {
            return fieldConfig;
        }
        var _errorBuilderOptions_defaultResultOptions;
        const { name: getResultName = defaultGetResultName , ...defaultResultOptions } = (_errorBuilderOptions_defaultResultOptions = errorBuilderOptions === null || errorBuilderOptions === void 0 ? void 0 : errorBuilderOptions.defaultResultOptions) !== null && _errorBuilderOptions_defaultResultOptions !== void 0 ? _errorBuilderOptions_defaultResultOptions : {
            name: defaultGetResultName
        };
        var _errorBuilderOptions_defaultUnionOptions;
        const { name: getUnionName = defaultGetUnionName , ...defaultUnionOptions } = (_errorBuilderOptions_defaultUnionOptions = errorBuilderOptions === null || errorBuilderOptions === void 0 ? void 0 : errorBuilderOptions.defaultUnionOptions) !== null && _errorBuilderOptions_defaultUnionOptions !== void 0 ? _errorBuilderOptions_defaultUnionOptions : {
            name: defaultGetUnionName
        };
        const parentTypeName = this.buildCache.getTypeConfig(fieldConfig.parentType).name;
        const { types =[] , result: { name: resultName = getResultName({
            parentTypeName,
            fieldName: capitalize(fieldConfig.name)
        }) , fields: resultFieldOptions , ...resultObjectOptions } = {} , union: { name: unionName = getUnionName({
            parentTypeName,
            fieldName: capitalize(fieldConfig.name)
        }) , ...unionOptions } = {} , dataField: { name: dataFieldName = 'data' , ...dataField } = {}  } = errorOptions;
        var _errorBuilderOptions_defaultTypes;
        const errorTypes = (0, _core.sortClasses)([
            ...new Set([
                ...types,
                ...(_errorBuilderOptions_defaultTypes = errorBuilderOptions === null || errorBuilderOptions === void 0 ? void 0 : errorBuilderOptions.defaultTypes) !== null && _errorBuilderOptions_defaultTypes !== void 0 ? _errorBuilderOptions_defaultTypes : []
            ])
        ]);
        var _errorOptions_directResult, _ref;
        const directResult = (_ref = (_errorOptions_directResult = errorOptions.directResult) !== null && _errorOptions_directResult !== void 0 ? _errorOptions_directResult : errorBuilderOptions === null || errorBuilderOptions === void 0 ? void 0 : errorBuilderOptions.directResult) !== null && _ref !== void 0 ? _ref : false;
        const typeRef = (0, _core.unwrapOutputFieldType)(fieldConfig.type);
        const typeName = this.builder.configStore.getTypeConfig(typeRef).name;
        const unionType = this.runUnique(resultName, ()=>{
            var _this_buildCache_getTypeConfig_extensions;
            let resultType;
            if (directResult && !Array.isArray(fieldConfig.pothosOptions.type)) {
                resultType = fieldConfig.pothosOptions.type;
                const resultConfig = this.builder.configStore.getTypeConfig(resultType);
                if (resultConfig.graphqlKind !== 'Object') {
                    throw new _core.PothosSchemaError(`Field ${parentTypeName}.${fieldConfig.name} must return an ObjectType when 'directResult' is set to true`);
                }
            } else {
                resultType = this.builder.objectRef(resultName);
                resultType.implement({
                    ...defaultResultOptions,
                    ...resultObjectOptions,
                    fields: (t)=>{
                        return {
                            ...resultFieldOptions === null || resultFieldOptions === void 0 ? void 0 : resultFieldOptions(t),
                            [dataFieldName]: t.field({
                                ...dataField,
                                type: fieldConfig.pothosOptions.type,
                                nullable: fieldConfig.type.kind === 'List' ? {
                                    items: fieldConfig.type.type.nullable,
                                    list: false
                                } : false,
                                resolve: (data)=>data
                            })
                        };
                    }
                });
            }
            const getDataloader = (_this_buildCache_getTypeConfig_extensions = this.buildCache.getTypeConfig((0, _core.unwrapOutputFieldType)(fieldConfig.type)).extensions) === null || _this_buildCache_getTypeConfig_extensions === void 0 ? void 0 : _this_buildCache_getTypeConfig_extensions.getDataloader;
            var _errorTypeMap_get;
            return this.builder.unionType(unionName, {
                types: [
                    ...errorTypes,
                    resultType
                ],
                resolveType: (obj)=>(_errorTypeMap_get = errorTypeMap.get(obj)) !== null && _errorTypeMap_get !== void 0 ? _errorTypeMap_get : resultType,
                ...defaultUnionOptions,
                ...unionOptions,
                extensions: {
                    ...unionOptions.extensions,
                    getDataloader,
                    pothosPrismaIndirectInclude: {
                        getType: ()=>typeName,
                        path: directResult ? [] : [
                            {
                                type: resultName,
                                name: dataFieldName
                            }
                        ]
                    }
                }
            });
        });
        return {
            ...fieldConfig,
            extensions: {
                ...fieldConfig.extensions,
                pothosErrors: errorTypes
            },
            type: {
                kind: 'Union',
                ref: unionType,
                nullable: fieldConfig.type.nullable
            }
        };
    }
    wrapResolve(resolver, fieldConfig) {
        var _fieldConfig_extensions;
        const pothosErrors = (_fieldConfig_extensions = fieldConfig.extensions) === null || _fieldConfig_extensions === void 0 ? void 0 : _fieldConfig_extensions.pothosErrors;
        if (!pothosErrors) {
            return resolver;
        }
        return async (...args)=>{
            try {
                return await resolver(...args);
            } catch (error) {
                for (const errorType of pothosErrors){
                    if (error instanceof errorType) {
                        const result = createErrorProxy(error, errorType, {
                            wrapped: true
                        });
                        errorTypeMap.set(result, errorType);
                        return result;
                    }
                }
                throw error;
            }
        };
    }
}
_core.default.registerPlugin(pluginName, PothosErrorsPlugin);
